emote/extra/onnx_storage.py

```diff
@@ -0,0 +1,164 @@
+
+    This is a mixin class that can be used with diamond inheritance to
+    create a specialized ONNX exporter.
+
+    Parameters:
+    :param directory: path to the directory where the files should be created. If it does not exist
+                      it will be created.
+    :param prefix: all file names will have this prefix.
+    """
+
+    def __init__(self, *args, directory: str, prefix: str = "savedmodel_", **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.storage = CRUDStorage(directory, prefix, extension="onnx")
+        self.queued_exports = Queue()
+        self.export_counter = 0
+
+        # Cache the version tag on startup.
+        # It takes about 20ms to calculate
+        self.version_tag = _get_version()
+        self.scopes = BlockTimers()
+
+    @abstractmethod
+    def generate_onnx(self) -> ModelProto:
+        """Generate a ModelProto for the policy."""
+        ...
+
+    def process_pending_exports(self):
+        """
+        If you are using `export_threadsafe` the main thread must call
+        this method regularly to make sure things are actually exported.
+        """
+        while self.queued_exports.qsize() > 0:
+            try:
+                item = self.queued_exports.get_nowait()
+            except Empty:
+                return
+            item.process(self)
+
+    def _export_onnx(self, metadata: Optional[Mapping[str, str]]) -> StorageItem:
+        def save_inner(export_path: str):
+            with self.scopes.scope("save"):
+                t0 = time.time()
+                model_proto = self.generate_onnx()
+                model_version = self.export_counter
+                self.export_counter += 1
+
+                model_proto.producer_name = "emote"
+                model_proto.domain = "dev.embark.ml"
+                model_proto.producer_version = self.version_tag
+                model_proto.model_version = model_version
+                model_proto.doc_string = "exported via Emote checkpointer"
+
+                if metadata is not None:
+                    helper.set_model_props(model_proto, metadata)
+
+                _save_protobuf(export_path, model_proto)
+
+                t1 = time.time()
+                logging.info(f"Exported onnx in {round((t1-t0)*1000)} ms")
+
+        with self.scopes.scope("create"):
+            return self.storage.create_with_saver(save_inner)
+
+    def _export(self, metadata: Optional[Mapping[str, str]], sync: bool) -> StorageItem:
+        # The actual onnx export needs to be done on the main thread.
+        item = QueuedExport(metadata)
+        self.queued_exports.put(item)
+        if sync:
+            # This will cause block_until_complete to never block
+            # because the work will have been completed already.
+            self.process_pending_exports()
+
+        return item.block_until_complete()
+
+    def export_threadsafe(self, metadata=None) -> StorageItem:
```

Comment:
This is never used. Is this whole API necessary when we in the code only follow the non-threadsafe path? We could cut a whole lot of code if this is not needed.